src/lando/headless_api/models/automation_job.py

```diff
@@ -0,0 +1,149 @@
+    priority = models.IntegerField(default=0)
+
+    # Duration of job from start to finish
+    duration_seconds = models.IntegerField(default=0)
+
+    # Reference to the target repo.
+    target_repo = models.ForeignKey(Repo, on_delete=models.SET_NULL, null=True)
+
+    @contextmanager
+    def processing(self):
+        """Mutex-like context manager that manages job processing miscellany.
+
+        This context manager facilitates graceful worker shutdown and
+        tracks the duration of the current job.
+        """
+        start_time = datetime.now()
+        try:
+            yield
+        finally:
+            self.duration_seconds = (datetime.now() - start_time).seconds
+
+    def to_api_status(self) -> dict[str, Any]:
+        """Return the job details as API status JSON."""
+        return {
+            "job_id": self.id,
+            "status_url": f"{settings.SITE_URL}/api/job/{self.id}",
+            "message": f"Job is in the {self.status} state.",
+            "created_at": self.created_at,
+            "status": self.status,
+        }
+
+    def transition_status(
+        self,
+        action: JobAction,
+        **kwargs,
+    ):
+        """Change the status and other applicable fields according to actions.
+
+        Args:
+            action (JobAction): the action to take, e.g. "land" or "fail"
+            **kwargs:
+                Additional arguments required by each action, e.g. `message` or
+                `commit_id`.
+        """
+        actions = {
+            JobAction.LAND: {
+                "required_params": ["commit_id"],
+                "status": JobStatus.LANDED,
+            },
+            JobAction.FAIL: {
+                "required_params": ["message"],
+                "status": JobStatus.FAILED,
+            },
+            JobAction.DEFER: {
+                "required_params": ["message"],
+                "status": JobStatus.DEFERRED,
+            },
+            JobAction.CANCEL: {
+                "required_params": [],
+                "status": JobStatus.CANCELLED,
+            },
+        }
+
+        if action not in actions:
+            raise ValueError(f"{action} is not a valid action")
+
+        required_params = actions[action]["required_params"]
+        if sorted(required_params) != sorted(kwargs.keys()):
+            missing_params = required_params - kwargs.keys()
+            raise ValueError(f"Missing {missing_params} params")
+
+        self.status = actions[action]["status"]
+
+        if action in (JobAction.FAIL, JobAction.DEFER):
+            self.error = kwargs["message"]
+
+        if action == JobAction.LAND:
+            self.landed_commit_id = kwargs["commit_id"]
+
+        self.save()
+
+    @classmethod
+    def next_job(cls, repositories: Optional[Iterable[str]] = None) -> Self:
+        """Return the next automation job."""
+        return (
+            cls.objects.filter(status__in=(JobStatus.SUBMITTED, JobStatus.DEFERRED))
```

Comment:
Question for my understanding: DEFERRED jobs don't look like they are treated any differently from the SUBMITTED job (e.g., no delay). Does this mean we may spin for while picking up and deferring the same job over and over again?

Reply:
Yes, but jobs are usually deferred due to tree closures, issues with pushes due to hgmo etc. Deferred usually means "this is safe to retry at some other time".